File: backend/src/main/java/com/carnalysys/util/EmployeeAvailability.java
package com.carnalysys.util;

import com.carnalysys.domain.OrderStatus;
import java.util.EnumSet;
import java.util.Set;

/** Normalizes workforce availability for API responses and assignment rules. */
public final class EmployeeAvailability {

  private static final Set<OrderStatus> ACTIVE_DELIVERY_ORDER_STATUSES =
      EnumSet.of(
          OrderStatus.placed,
          OrderStatus.confirmed,
          OrderStatus.processing,
          OrderStatus.shipped);

  private EmployeeAvailability() {}

  public static boolean isActiveDeliveryOrderStatus(OrderStatus status) {
    return status != null && ACTIVE_DELIVERY_ORDER_STATUSES.contains(status);
  }

  /**
   * Effective label for UI: offline only when stored status is offline; busy when stored busy or
   * active assigned orders exist; online for online/free when not busy.
   */
  public static String effectiveStatus(String availabilityStatus, boolean hasActiveAssignedOrders) {
    String raw =
        availabilityStatus == null || availabilityStatus.isBlank()
            ? "offline"
            : availabilityStatus.trim().toLowerCase();
    if ("offline".equals(raw)) {
      return "offline";
    }
    if (hasActiveAssignedOrders) {
      return "busy";
    }
    if ("online".equals(raw) || "free".equals(raw) || "busy".equals(raw)) {
      return "online";
    }
    return "offline";
  }

  public static boolean isStoredOnline(String availabilityStatus) {
    String raw =
        availabilityStatus == null ? "" : availabilityStatus.trim().toLowerCase();
    return "online".equals(raw) || "free".equals(raw);
  }
}

File: src/lib/employeeAvailability.js
/**
 * Effective availability from an employee API row (`availability` only — not onboarding `status`).
 * @param {{ availability?: string | null } | null | undefined} row
 * @returns {'online' | 'busy' | 'offline'}
 */
export function employeeAvailabilityFromRow(row) {
  return normalizeEmployeeAvailability(row?.availability)
}

/**
 * Workforce availability display (matches backend EmployeeAvailability.effectiveStatus).
 * @param {string | null | undefined} raw
 * @returns {'online' | 'busy' | 'offline'}
 */
export function normalizeEmployeeAvailability(raw) {
  const s = String(raw ?? '').trim().toLowerCase()
  if (s === 'offline') return 'offline'
  if (s === 'busy') return 'busy'
  if (s === 'online' || s === 'free') return 'online'
  return 'offline'
}

/** @param {string | null | undefined} raw */
export function employeeAvailabilityLabel(raw) {
  const status = normalizeEmployeeAvailability(raw)
  if (status === 'online') return 'Online'
  if (status === 'busy') return 'Busy'
  return 'Offline'
}

/** Uppercase label for compact UI (orders dropdown). */
export function employeeAvailabilityShortLabel(raw) {
  return employeeAvailabilityLabel(raw).toUpperCase()
}

/** @param {string | null | undefined} raw */
export function isEmployeeOnlineForAssignment(raw) {
  return normalizeEmployeeAvailability(raw) === 'online'
}
